Hex viewer

davedelong opened the pull request:
This adds a hex viewer to Pasteboard Viewer and fixes #24.

This uses the [HexFiend framework](https://github.com/hexfiend/hexfiend) to add a hex viewer to the app, which can be accessed by toggling a picker in the toolbar.

Some things to note:

- This depends on the `package` branch of the HexFiend project, as that is the branch that adds a Package.swift file. It is technically unreleased, although it appears to be fully functional.
- The format of the "extra info" in the toolbar is meant to mimic the format shown in the Hex Fiend app as closely as possible.
- Selection is not preserved when toggling between text and hex modes.

<img width="989" alt="Screenshot 2024-05-02 at 9 10 09 PM" src="https://github.com/sindresorhus/Pasteboard-Viewer/assets/112699/d9f4eb42-4f0f-4e60-bc54-96ffb3b25199">
<img width="989" alt="Screenshot 2024-05-02 at 9 10 21 PM" src="https://github.com/sindresorhus/Pasteboard-Viewer/assets/112699/97625980-7d57-4a55-9f69-71a83a17f291">

davedelong committed:
Hex viewer

sindresorhus commented on Pasteboard Viewer/ContentsScreen.swift:
```diff
@@ -258,3 +320,67 @@ extension Pasteboard.Type_ {
 }
+
+struct HexView: NSViewRepresentable {
+	class Coordinator: HFRepresenter, ObservableObject {
```
```suggestion
	class Coordinator: HFRepresenter {
```

It doesn't look like it's used for anything.

sindresorhus commented on Pasteboard Viewer/ContentsScreen.swift:
```diff
@@ -258,3 +320,67 @@ extension Pasteboard.Type_ {
+			guard isUpdating == false else {
+				return
+			}
+			if bits.contains(.contentValue) || bits.contains(.selectedRanges) {
```
I would use guard here to reduce nesting.

sindresorhus commented:
Thanks for working on this! This is already looking pretty good.

- You need to if-def it out for iOS.
- The selection needs to be reset when the pasteboard changes.
	- I also think it makes sense to reset the selection when switching between pasteboard items.
- I noticed it's quite slow to select something when viewing an image. Not sure if the problem is the framework or the wrapper code.
- Would be useful to have the text/hex picker in the "View" menu bar menu too, with keyboard shortcuts. But not a requirement for this pull request.

sindresorhus committed:
Update ContentsScreen.swift

sindresorhus commented on Pasteboard Viewer/ContentsScreen.swift:
```diff
@@ -1,12 +1,66 @@
+	}
+
+	private var hexBody: some View {
+		// this uses a .number formatter instead of .byteCount,
+		// because a byte count format doesn't have a way to exclude the grouping character
+		let count = (type.data()?.count ?? 0)
+		let length = count.formatted(.number.grouping(.never)) + " " + (count == 1 ? "byte" : "bytes")
+		let extraInfo: String
+		if selectedByteRanges.isEmpty {
+			// no selection
+			extraInfo = length
+		} else if let range = selectedByteRanges.first, selectedByteRanges.count == 1 {
+			// single or empty selection
+			let offset = range.lowerBound.formatted(.number.grouping(.never))
+			if range.lowerBound == range.upperBound {
+				// no selection
+				extraInfo = length
+			} else if range.upperBound - range.lowerBound == 1 {
+				// single byte
+				extraInfo = "Byte \(offset) selected out of \(length)"
+			} else {
+				// multiple bytes
+				let count = (range.upperBound - range.lowerBound).formatted(.number.grouping(.never))
+				extraInfo = "\(count) bytes selected at offset \(offset) out of \(length)"
+			}
+		} else {
+			// multiple selection
+			let totalCount = selectedByteRanges.reduce(into: UInt64.zero, { $0 += ($1.upperBound - $1.lowerBound) })
+				.formatted(.number.grouping(.never))
+
+			extraInfo = "\(totalCount) selected at multiple offsets out of \(length)"
+		}
```
I would move this to a `hexBodyExtraInfo` getter to improve readability as then each branch could use `return`. And it could also use `guard`.

davedelong committed:
Add View menu shortcuts (⌘1 and ⌘2)

davedelong committed:
Address most review feedback

- macOS only
- extract extra into to helper property
- appease the SwiftLint gods
- simplify HexView.Coordinator

davedelong committed:
Use `@Default` instead of `@AppStorage`

davedelong committed:
Force state to reset

When either the selected type changes or there's new data on the pasteboard, force the contents view to reset, thereby losing selection state

davedelong commented:
> * I noticed it's quite slow to select something when viewing an image. Not sure if the problem is the framework or the wrapper code.

I took a quick look at this in instruments, and from what I can tell there's an identical spike when showing the text view as showing the hex view. I'm reasonably confident that this is down to retrieving the `type.data()` from the pasteboard to provide to the UI. There's probably an optimization that could be made there around caching it once it's been read, so that subsequent calls to `.data()` are instantaneous.

davedelong commented:
Besides the hiccup when showing large datas (like images), all the other feedback has been addressed, including adding the ⌘1 and ⌘2 switcher in the View menu. 😃

davedelong committed:
Merge branch 'main' into hex-viewer

davedelong commented:
(I also confirmed that the branch builds cleanly for iOS)

sindresorhus commented:
Looks great 🙌

sindresorhus commented:
Update submitted.
